### PR TITLE
Add `vue/require-default-export` rule

### LINKED ISSUE #2383: Require default export on `<script>`
**Please describe what the rule should do:**

To warn when there's a `<script>` tag with no default export.

With the new `<script setup>` syntax, I've noticed cases where devs forget or miss the `setup` and can't figure why their code is not working correctly.

**What category should the rule belong to?**

[ ] Enforces code style (layout)
[x] Warns about a potential error (problem)
[ ] Suggests an alternate way of doing something (suggestion)
[ ] Other (please specify:)

**Provide 2-3 code examples that this rule should warn about:**

```vue
<!-- Forgetting to add `setup` -->
<script>
const value = 1;
</script>
```

**Additional context**

From: https://github.com/vuejs/eslint-plugin-vue/issues/2328

### PR DESCRIPTION
**Pull request by ItMaga**
closes #2383

### CODE REVIEW

File: tests/lib/rules/require-default-export.js
[the remark this answers is not in the document]
Comment (ItMaga) [reply tests/lib/rules/require-default-export.js:1]: Sorry, I'm not sure that I got it correctly

> `.js` files with components → should be reported or ignored

Do you mean, that in the `.js` file, for example:
```javascript
import { defineComponent } from 'vue';

defineComponent({});
```
should be reported? I think we should always ignore `.js` files

File: lib/rules/require-default-export.js [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+'use strict'
+
+const utils = require('../utils')
+
+module.exports = {
+  meta: {
+    type: 'problem',
+    docs: {
+      description: 'require default export',
+      categories: ['vue3-essential', 'vue2-essential'],
```
[the remark this answers is not in the document]
Comment (ItMaga) [reply lib/rules/require-default-export.js:14]: I was thinking, should I add or don't, so now I removed it

File: docs/rules/object-curly-newline.md [hunk 2]
```diff
@@ -36,4 +36,4 @@ This rule was introduced in eslint-plugin-vue v7.0.0
 - [Rule source](https://github.com/vuejs/eslint-plugin-vue/blob/master/lib/rules/object-curly-newline.js)
 - [Test source](https://github.com/vuejs/eslint-plugin-vue/blob/master/tests/lib/rules/object-curly-newline.js)
 
-<sup>Taken with ❤️ [from ESLint Stylistic](https://eslint.style/rules/js/object-curly-newline)</sup>
+<sup>Taken with ❤️ [from ESLint Stylistic](https://eslint.style/rules/ts/object-curly-newline)</sup>
```
[the remark this answers is not in the document]
Comment (ItMaga) [reply docs/rules/object-curly-newline.md:39]: I tried to run `npm install` and then `npm run update`, but it didn't help. So I reverted manually.

File: docs/rules/require-default-export.md
[the remark this answers is not in the document]
Comment (FloEdelmann) [reply docs/rules/require-default-export.md:1]: Please also add `vue/require-default-export` as a related rule to `vue/one-component-per-file` now.

File: tests/lib/rules/require-default-export.js
[the remark this answers is not in the document]
Comment (FloEdelmann) [reply tests/lib/rules/require-default-export.js:1]: > I think we should always ignore `.js` files

That's fine for me, too, I just noted that it should be tested. But you added them in [`77f1126`](https://github.com/vuejs/eslint-plugin-vue/pull/2494/commits/77f112642480065555d7f37bc732320e18a54feb), so it's fine.

File: tests/lib/rules/require-default-export.js [hunk 3]
```diff
@@ -0,0 +1,156 @@
...
+      filename: 'test.vue',
+      code: `
+      <script>
+      const foo = 'foo';
+      </script>
+      `,
+      errors: [
+        {
+          messageId: 'mustDefaultExport',
+          line: 1
```
Comment (FloEdelmann) [inline tests/lib/rules/require-default-export.js:90]: Can you change the reported range to be only the end `</script>` tag? That's usually where the default export goes. Also, please add `endLine`, `column` and `endColumn` to all test cases.

File: lib/rules/require-default-export.js [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+    type: 'problem',
+    docs: {
+      description: 'require components to be the default export',
+      categories: undefined,
+      url: 'https://eslint.vuejs.org/rules/require-default-export.html'
+    },
+    fixable: null,
+    schema: [],
+    messages: {
+      mustDefaultExport: 'The component must be the default export.'
```
Comment (FloEdelmann) [inline lib/rules/require-default-export.js:20]: Now the message is not really fitting for when there is no export at all. I think we either need different error messages, or a single refined one:

```suggestion
      mustDefaultExport: 'Script block should have component as default export.'
```

or:

1. If there is no export at all and no component (`defineComponent(…)` or `Vue.component(…)` or similar) in a `.vue` file: `Missing default export.`
2. If there is a component: `Component must be the default export.`

**Review by FloEdelmann — CHANGES_REQUESTED** [review]

File: lib/rules/require-default-export.js (hunk 4 above)
Comment (ItMaga) [reply lib/rules/require-default-export.js]: Thank you for the clarification. Fixed

**Review by FloEdelmann — APPROVED** [review]
Nice, looks good to me now :slightly_smiling_face:

**Review by ota-meshi — APPROVED** [review]
LGTM! Thank you!

**Review by SamyDgz — APPROVED** [review]

**Pull request merged**